Add `Monitoring` section into RuntimeClass KEP (#756)
kubernetes/enhancements: keps/sig-node/runtime-class.md

the `RuntimeHandler` is not recognized by the CRI implementation, then `RunPodSandbox` will return
an error.

#### Monitoring

The first round of monitoring implementation for `RuntimeClass` will cover the
following three areas:

- `how popular every runtime is?` This can be done by adding a new label into

>   [kubelet\_running\_pod\_count](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/metrics/metrics.go#L256)

Reviewer: nit: Links to specific lines should always pin a commit, otherwise the line may drift. Just press 'y' on github to replace the URL with a pinned version.

- `how robust every runtime is?` This can be done by adding a new metric to
  track the errors of different RuntimeClass of
  [RunPodSandbox](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/instrumented_services.go#L179)
  operations.
- `how expensive every runtime is in terms of time overhead?` This can be done
  by adding a new metric to track the latencies of different RuntimeClass of
  [RunPodSandbox](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/instrumented_services.go#L179)
  operations.

The following monitoring areas will be skipped for now, and be considered after
the RuntimeClass scheduling is implemented:

- how many runtimes a cluster supports?
- how many scheduling failures caused by unsupported runtimes or insufficient
  resources of a certain runtime?

Currently, we assume that all the nodes in a cluster are homogeneous. After
heterogeneous clusters are implemented, we may need to monitor how many runtimes
a node supports.

[runpodsandbox]: https://github.com/kubernetes/kubernetes/blob/b05a61e299777c2030fbcf27a396aff21b35f01b/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto#L344

> ### Risks and Mitigations

Reviewer: Can you delete `Details TBD.` on [line 361](https://github.com/kubernetes/enhancements/pull/756/files#diff-d9093bb5b949f3a78d08184362eb3d48R361)

[runpodsandbox]: https://github.com/kubernetes/kubernetes/blob/b05a61e299777c2030fbcf27a396aff21b35f01b/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto#L344

#### Monitoring

The first round of monitoring implementation for `RuntimeClass` covers the

> following two areas and has finished (tracked in

Reviewer: nit
```suggestion
following two areas and is finished (tracked in
```

[#73058](https://github.com/kubernetes/kubernetes/issues/73058)):

> - `how robust every runtime is?` A new metric

Reviewer: nit
```suggestion
- `how robust is every runtime?` A new metric
```

  [RunPodSandboxErrors](https://github.com/kubernetes/kubernetes/blob/596a48dd64bcaa01c1d2515dc79a558a4466d463/pkg/kubelet/metrics/metrics.go#L351)
  is added to track the RunPodSandbox operation errors, broken down by
  RuntimeClass.

> - `how expensive every runtime is in terms of time overhead?` A new metric

Reviewer: nit
```suggestion
- `how expensive is every runtime in terms of latency?` A new metric
```

  [RunPodSandboxDuration](https://github.com/kubernetes/kubernetes/blob/596a48dd64bcaa01c1d2515dc79a558a4466d463/pkg/kubelet/metrics/metrics.go#L341)
  is added to track the duration of RunPodSandbox operations, broken down by
  RuntimeClass.

> The following monitoring areas will be skipped for now, and be considered after

Reviewer: nit
```suggestion
The following monitoring areas will be skipped for now, but may be considered after
```

the RuntimeClass scheduling is implemented:

> - how many runtimes a cluster supports?

Reviewer: ```suggestion
- how many runtimes does a cluster support?
```

> - how many scheduling failures caused by unsupported runtimes or insufficient

Reviewer: ```suggestion
- how many scheduling failures were caused by unsupported runtimes or insufficient
```

  - [ ] [CRI validation tests][cri-validation]
  - [ ] RuntimeClasses are configured in the E2E environment with test coverage of a non-default
    RuntimeClass

> - [ ] Comprehensive coverage of RuntimeClass metrics. [#73058](http://issue.k8s.io/73058)

Reviewer: Check the box :)
```suggestion
- [x] Comprehensive coverage of RuntimeClass metrics. [#73058](http://issue.k8s.io/73058)
```
